From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Thu, 1 Dec 2022 14:18:46 +0100
Subject: [PATCH] Cache world generator sea level

License: LGPL-3.0-only (https://www.gnu.org/licenses/lgpl-3.0.html)
Gale - https://galemc.org

* Lithium description *

Use cached sea level instead of retrieving from the registry every time.
This method is called for every block in the chunk so this will save a lot of registry lookups.
Lazy initialization added to avoid unbound registry value crash

This patch is based on the following mixin:
"net/caffeinemc/mods/lithium/mixin/gen/cached_generator_settings/NoiseBasedChunkGeneratorMixin.java"
By: SuperCoder7979 <k.pranav@gmail.com>, 2No2Name <2No2Name@web.de>
As part of: Lithium (https://github.com/CaffeineMC/lithium)
Licensed under: LGPL-3.0-only (https://www.gnu.org/licenses/lgpl-3.0.html)

// SYNC SOURCE START
rev: 01597fb879b7a6d17c2420ec4e25a98df3bc8e7b
branch: develop
repo: https://github.com/CaffeineMC/lithium.git
track:
- "common/src/main/java/net/caffeinemc/mods/lithium/mixin/gen/cached_generator_settings/NoiseBasedChunkGeneratorMixin.java"
// SYNC SOURCE END

diff --git a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
index f0529d233b791e226cec469b2efebcffd02c6d58..ff328b5435e4e5ca148680d912078b939927aa79 100644
--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -69,6 +69,7 @@ public final class NoiseBasedChunkGenerator extends ChunkGenerator {
     private static final DecimalFormat DEBUG_DENSITY_FORMAT = new DecimalFormat("0.000", DecimalFormatSymbols.getInstance(Locale.ROOT));
     private final Holder<NoiseGeneratorSettings> settings;
     private final Supplier<Aquifer.FluidPicker> globalFluidPicker;
+    private int cachedSeaLevel = Integer.MIN_VALUE; // Gale - Lithium - cache world generator sea level
 
     public NoiseBasedChunkGenerator(final BiomeSource biomeSource, final Holder<NoiseGeneratorSettings> settings) {
         super(biomeSource);
@@ -500,7 +501,12 @@ public final class NoiseBasedChunkGenerator extends ChunkGenerator {
 
     @Override
     public int getSeaLevel() {
-        return this.settings.value().seaLevel();
+        // Gale start - Lithium - cache world generator sea level
+        if (cachedSeaLevel == Integer.MIN_VALUE) {
+            this.cachedSeaLevel = this.settings.value().seaLevel();
+        }
+        return this.cachedSeaLevel;
+        // Gale end - Lithium - cache world generator sea level
     }
 
     @Override
